src/lib.rs

```diff
@@ -17,244 +16,178 @@
-    Truncated {
-        /// The maximum size of the subset is 2^num_bits.
-        num_bits: usize,
-    },
+    /// Sample field elements from a subset of the field, specified by the maximum number of bits.
+    Truncated(usize),
+}
+
+impl FieldElementSize {
+    pub(crate) fn num_bits<F: PrimeField>(&self) -> usize {
+        if let FieldElementSize::Truncated(num_bits) = self {
+            if *num_bits > (F::Params::MODULUS_BITS as usize) {
+                panic!("num_bits is greater than the capacity of the field.")
+            }
+            *num_bits.min(&(F::Params::CAPACITY as usize))
+        } else {
+            F::Params::CAPACITY as usize
+        }
+    }
+
+    /// Calculate the sum of field element sizes in `elements`.
+    pub fn sum<F: PrimeField>(elements: &[Self]) -> usize {
+        elements.iter().map(|item| item.num_bits::<F>()).sum()
+    }
 }
 
 /// The interface for a cryptographic sponge.
 /// A sponge can `absorb` or take in inputs and later `squeeze` or output bytes or field elements.
 /// The outputs are dependent on previous `absorb` and `squeeze` calls.
-pub trait CryptographicSponge<F: PrimeField> {
+pub trait CryptographicSponge: Clone {
+    /// Parameters used by the sponge.
+    type Parameters;
+
     /// Initialize a new instance of the sponge.
-    fn new() -> Self;
+    fn new(params: &Self::Parameters) -> Self;
 
     /// Absorb an input into the sponge.
-    fn absorb(&mut self, input: &impl Absorbable<F>);
+    fn absorb(&mut self, input: &impl Absorb);
 
     /// Squeeze `num_bytes` bytes from the sponge.
     fn squeeze_bytes(&mut self, num_bytes: usize) -> Vec<u8>;
 
+    /// Squeeze `num_bits` bits from the sponge.
+    fn squeeze_bits(&mut self, num_bits: usize) -> Vec<bool>;
+
     /// Squeeze `sizes.len()` field elements from the sponge, where the `i`-th element of
     /// the output has size `sizes[i]`.
-    fn squeeze_field_elements_with_sizes(&mut self, sizes: &[FieldElementSize]) -> Vec<F>;
-
-    /// Squeeze `num_elements` field elements from the sponge.
-    fn squeeze_field_elements(&mut self, num_elements: usize) -> Vec<F> {
-        self.squeeze_field_elements_with_sizes(
-            vec![FieldElementSize::Full; num_elements].as_slice(),
-        )
-    }
-}
-
-/// An interface for objects that can be absorbed by a `CryptographicSponge`.
-pub trait Absorbable<F: PrimeField> {
-    /// Converts the object into a list of bytes that can be absorbed by a `CryptographicSponge`.
-    fn to_sponge_bytes(&self) -> Vec<u8>;
+    ///
+    /// If the implementation is field-based, to squeeze native field elements,
+    /// call `self.squeeze_native_field_elements` instead.
+    ///
+    /// TODO: Support general Field.
+    ///
+    /// Note that when `FieldElementSize` is `FULL`, the output is not strictly uniform. Output
+    /// space is uniform in \[0, 2^{F::MODULUS_BITS - 1}\]
+    fn squeeze_field_elements_with_sizes<F: PrimeField>(
```

Comment:
This makes gadgets implementations unnecessarily inefficient, no? It requires converting to bits even for `FieldElementSize::Full`.

Reply:
fwd @Will-Lin4 : not exactly sure how to do here. May be in specific implementation, we overide this method by checking if field is native and size is `Full`, then we call `squeeze_native_field_elements` (otherwise use default implementation)?

Reply:
@Pratyush: I think the code for this function was originally used for squeezing out nonnative field elements, so it didn't take into account the possibility that F is the native field.

@tsunrise: The override approach you suggest sounds good to me.

Reply:
@Will-Lin4 What shall we do if we want to squeeze native field when size is not `Full`? Shall we convert it to bytes?

Reply:
fixed (if assume we convert native field to bytes if we do not squeeze `Full` size)